### PR TITLE
Add Dockerfile for Oasis Contract Watcher

### PR DESCRIPTION
**Pull request by rmulhol**
The only departure from our usual pattern is that I inlined the startup command into the Dockerfile, instead of putting it in a `startup_script.sh`. This has the downside of removing validation for required env vars, but reduces clutter a bit. I was also running into a permissioning issue executing `startup_script.sh` from within the container 🤔

### CODE REVIEW

File: dockerfiles/README.md [hunk 1]
```diff
@@ -1,15 +1,35 @@
...
 ```
-docker run -e DATABASE_USER="user" -e DATABASE_PASSWORD="pw" -e DATABASE_HOSTNAME="host" -e DATABASE_PORT="port" -e DATABASE_NAME="name" -e STARTING_BLOCK_NUMBER=0 -e CLIENT_IPCPATH="path" -t header_sync:latest
+docker run -e DATABASE_USER=user -e DATABASE_PASSWORD=password -e DATABASE_HOSTNAME=host -e DATABASE_PORT=port -e DATABASE_NAME=name -e STARTING_BLOCK_NUMBER=0 -e CLIENT_IPCPATH=path -it header_sync:latest
 ```
+Note: must replace env var values with appropriate replacements given your database/node setup.
+
+## contractWatcher
+Sync events from given contract(s).
+
+Note: depends on separate headerSync process persisting headers to same database.
```
Comment (elizabethengelman) [inline dockerfiles/README.md]: I wonder if would be good to explicitly say that `headerSync` needs to be run before and/or in parallel with `contractWatcher`.

File: dockerfiles/README.md [hunk 2]
```diff
@@ -1,15 +1,35 @@
...
+#### Build
+From project root directory:
+```
+docker build -f dockerfiles/contract_watcher/oasis/Dockerfile . -t contract_watcher:oasis-latest
+```
+
+#### Run
+```
+docker run -e DATABASE_USER=user -e DATABASE_PASSWORD=password -e DATABASE_HOSTNAME=host -e DATABASE_PORT=port -e DATABASE_NAME=name -e CLIENT_IPCPATH=path -it contract_watcher:oasis-latest
+```
```
Comment (elizabethengelman) [inline dockerfiles/README.md]: Not sure how necessary it is, but do you think we should include a note about using `host.docker.internal` to access localhost on MacOS, and `--network="host" for linux?

**Comment by rmulhol** [comment]
> LGTM!
> 
> Once thing I am wondering about is if it makes sense to merge these changes into vdb-mcd-transformers (or a new repo) instead of this one, since it's specific to oasis.

I think I would lean toward merging to this repo or not at all, given that it requires the executable defined here and doesn't otherwise impact the MCD plugin. My thinking was that we could potentially populate the `dockerfiles/contract_watcher/` directory with example images for running the contract watcher, but I don't really have a good idea for how we'd bound what examples are worth merging 🤔

**Comment by elizabethengelman** [comment]
> I think I would lean toward merging to this repo or not at all, given that it requires the executable defined here and doesn't otherwise impact the MCD plugin. My thinking was that we could potentially populate the dockerfiles/contract_watcher/ directory with example images for running the contract watcher, but I don't really have a good idea for how we'd bound what examples are worth merging 🤔

That makes sense to me. Is the thought that a requirement for things that are worth merging are more general use cases?

**Comment by rmulhol** [comment]
> > I think I would lean toward merging to this repo or not at all, given that it requires the executable defined here and doesn't otherwise impact the MCD plugin. My thinking was that we could potentially populate the dockerfiles/contract_watcher/ directory with example images for running the contract watcher, but I don't really have a good idea for how we'd bound what examples are worth merging 🤔
> 
> That makes sense to me. Is the thought that a requirement for things that are worth merging are more general use cases?

That's a good question. I was thinking that every Dockerfile _usually_ serves a general use case, but the contract watcher would be an exception since it's always targeting a specific contract. In that thinking, the purpose of merging this would be more to serve as an example rather than a general use case. However, I could see a case for not merging this and instead creating a more general contract watcher Dockerfile where you pass in the address/abi/deployment block as a build arg or env var 🤔

**Comment by elizabethengelman** [comment]
> I could see a case for not merging this and instead creating a more general contract watcher Dockerfile where you pass in the address/abi/deployment block as a build arg or env var 🤔

That's a cool idea! I'm also good with merging this in, and then we can refactor how the docker image behaves in the future (e.g. allow it to take address, abi and deployment blocks in as args).

**Comment by rmulhol** [comment]
Going to close this and reconsider writing a more general contract watcher Dockerfile with address/abi configurable. Have come around to thinking we prob shouldn't merge something this specific without a story to prompt it (and perhaps corresponding CI changes to rebuild as new PRs are merged)

**Pull request closed without merging**